core/appmodule/v2/handlers.go

```diff
@@ -16,47 +17,152 @@ type (
 	PostMsgHandler = func(ctx context.Context, msg, msgResp transaction.Msg) error
 )
 
-// msg handler
-
+// PreMsgRouter is a router that allows you to register PreMsgHandlers for specific message types.
 type PreMsgRouter interface {
 	// RegisterPreHandler will register a specific message handler hooking into the message with
 	// the provided name.
-	RegisterPreHandler(msgName string, handler PreMsgHandler)
+	RegisterPreMsgHandler(msgName string, handler PreMsgHandler)
 	// RegisterGlobalPreHandler will register a global message handler hooking into any message
 	// being executed.
-	RegisterGlobalPreHandler(handler PreMsgHandler)
+	RegisterGlobalPreMsgHandler(handler PreMsgHandler)
 }
 
+// HasPreMsgHandlers is an interface that modules must implement if they want to register PreMsgHandlers.
 type HasPreMsgHandlers interface {
 	RegisterPreMsgHandlers(router PreMsgRouter)
 }
 
-type MsgRouter interface {
-	Register(msgName string, handler Handler)
-}
+// RegisterMsgPreHandler is a helper function that modules can use to not lose type safety when registering PreMsgHandler to the
+// PreMsgRouter. Example usage:
+// ```go
+//
+//	func (h Handlers) BeforeSend(ctx context.Context, req *types.MsgSend) (*types.QueryBalanceResponse, error) {
+//	      ... before send logic ...
+//	}
+//
+//	func (m Module) RegisterPreMsgHandlers(router appmodule.PreMsgRouter) {
+//		handlers := keeper.NewHandlers(m.keeper)
```

Comment:
a way  to use it. it's to simply define before send on the keeper and handle it via keeper, if we wanto to do it msg server style and have a wrapper like handlers then that is also fine.

Reply:
I like to think of the keeper as something only for state (basically, just holding public/private collections), and the business logic is somewhere else.